### ISSUE TITLE
Slow View Performance When retrieving 2000+ docs

### ISSUE THREAD

**Issue by jfspencer** [pouchdb/pouchdb#3468]
I am working on a project the needs to query a view that returns a couple thousand docs at a time.

The design doc is really simple, every doc in the db has the doc.key . I am letting the view build before I query it.
    var designDoc = {
    _id: '_design/view_name',
    views: { view_name: { map: function(doc) {emit(doc.key); }.toString() } }
    };

my query options are
    var options = {key: key , stale: 'update_after' , descending: true, include_docs: true};

then I am querying like so
    pouch.db.query('design_doc', options).then(function(){....});

when querying a key that will return approx. 2000 docs it takes about 5000ms - 6000ms to complete.

In comparison using allDocs + lodash filters takes about 116ms - 130ms for the same results.

Is this a limitation of views or are there other optimizations that I can make? I would just use allDocs, but the eventual size of the db will be in the 50,000-75,000 doc range. I assume allDocs performance is linear(maybe not?), in which case all docs would take about 1600ms if it had to sift through 50,000 docs.

I have read through the 12 pro tips blog post, and the writing views suggestions in the post announcing view support and optimized where possible. I am using the ionic Framework with the cordova SQLite plugin.

**Comment by nolanlawson** [comment]
Some suggestions:

1) Try it without the SQLite plugin. It may be faster. The SQLite plugin is basically only better at 1) not having an upper storage quota, and 2) running on a background thread as opposed to the UI thread. but because it has to pass messages as strings between Java and JavaScript, it's actually slower in a lot of cases.

2) Yes, `allDocs()` will always be faster than views. That's a limitation of PouchDB.

3) Try pouchdb master; we have some optimizations to view performance: http://pouchtest.com/nightly/index.html

4) Don't fetch 1,000 docs at a time, especially in the SQLite plugin. It has to basically serialize that out into a string and pass it from the Java to the JavaScript. Use pagination and play with the `limit`; I'm guessing the sweet spot is more like 100.

5) Don't pull out the entire doc; maybe your docs are huge? Try pulling out just a subset of the data. This could be either a value on your view, or it could be separate docs (maybe even in a separate database) that contain a "summary" of the document as opposed to the entire document.

Let me know how that goes. I am really interested in improving view performance. :)

**Comment by jfspencer** [comment]
@nolanlawson 
Thanks for the quick feedback! 

some feed back to your suggestions -
1/6) Haven't tried switching adapters yet, when I do I'll let you know how they compare.
2) makes sense
3) pouchdb master performed better than 3.2.1 - about 2-5% faster
4) Paging the queries made a big difference. Your right with the limit around 100, which dropped load time to ~2000ms. I didn't expect paging to have such a big effect. This was a good learning exercise to test the limits of large queries. I ended up making a another view with a complex key that permits me to gather the exact docs from a query which ended up being a lot faster than allDocs + filter.
5) each doc ranges from 5-15 keys with a few of them storing small arrays(less than 10 elements).

to sum up the tests:
allDocs + filter = 116-130ms (3500 docs with lodash.where resulting in 14 docs)
single key view + filter = 5500-6000ms (2200 docs with lodash.where resulting in 14 docs)
paged general complex key view + filter = 2000-2150ms (2200 docs with several lodash functions resulting in 14 docs)
targeted complex key view no filter = 21-28ms (query gets the 14 docs  - views done right for the win :)

**Comment by nolanlawson** [comment]
Awesome \o/. Are you still using the SQLite plugin, out of curiosity?

**Comment by jfspencer** [comment]
Yes all tests were done with the SQLite plugin.

On Friday, January 30, 2015, Nolan Lawson <notifications@github.com> wrote:

> Awesome \o/. Are you still using the SQLite plugin, out of curiosity?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pouchdb/pouchdb/issues/3468#issuecomment-72200052>.
>


-- 

Cheers,


James

**Comment by robertkeizer** [comment]
@jfspencer I got some performance gains by using views that emit `null` as the value, and then using an `allDocs` query with `keys` specified.

**Comment by nolanlawson** [comment]
Sounds like this got resolved; glad it worked out.

**Comment by Jars1991** [comment]
Hi, I am using alldocs with include_docs: true, i have 1928 elements and takes about 1178.5 ms, is it slow ?, i tested with include_docs: false and takes around 400ms, what can you tell me about this issue ?
